qiskit/transpiler/passes/optimization/collect_linear_functions.py

```diff
@@ -0,0 +1,84 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""Replace each sequence of CX and SWAP gates by a single LinearFunction gate."""
+
+from qiskit.circuit.library.standard_gates import CXGate, SwapGate
+from qiskit.circuit.library.generalized_gates import LinearFunction
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.circuit import QuantumCircuit
+
+
+class CollectLinearFunctions(TransformationPass):
+    """Collect blocks of linear gates (CX and SWAP) and replace them by linear functions."""
+
+    @staticmethod
+    def _is_linear_gate(op):
+        return isinstance(op, (CXGate, SwapGate)) and op.condition is None
+
+    # Called when reached the end of the linear block (either the next gate
+    # is not linear, or no more nodes)
+    @staticmethod
+    def _finalize_processing_block(cur_nodes, cur_qubits, blocks):
+        # Collect only blocks comprising at least 2 gates.
+        # ToDo: possibly make the minimum number of gates as a parameter
+        if len(cur_nodes) >= 2:
+            cur_wire_pos_map = dict((qb, ix) for ix, qb in enumerate(cur_qubits))
+            blocks.append((cur_nodes, cur_wire_pos_map))
+
+    # For now, we implement the naive greedy algorithm that makes a linear sweep over
+    # nodes in DAG (using the topological order) and collects blocks of linear gates.
```

Comment:
As a future enhancement, it would be good to extend `CollectMultiQBlocks` to be able to collect different blocks types beyond just unitaries, as it has some logic to try to maximize the size of the collected block, but I'll open a separate issue for this.